internal/planner/upsert.go

```diff
@@ -0,0 +1,185 @@
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package planner
+
+import (
+	"github.com/sourcenetwork/defradb/client"
+	"github.com/sourcenetwork/defradb/client/request"
+	"github.com/sourcenetwork/defradb/internal/core"
+	"github.com/sourcenetwork/defradb/internal/planner/mapper"
+)
+
+type upsertNode struct {
+	documentIterator
+	docMapper
+
+	p           *Planner
+	collection  client.Collection
+	filter      *mapper.Filter
+	createInput map[string]any
+	updateInput map[string]any
+	doc         *client.Document
+	isUpserting bool
+	results     planNode
+}
+
+// Next only returns once.
+func (n *upsertNode) Next() (bool, error) {
+	if n.isUpserting {
+		next, err := n.results.Next()
+		if err != nil {
+			return false, err
+		}
+
+		switch next {
+		case true:
+			n.currentValue = n.results.Value()
+			// make sure multiple documents do not match
+			next, err := n.results.Next()
+			if err != nil {
+				return false, err
+			}
+			if next {
+				return false, ErrUpsertMultipleDocuments
+			}
+			docID, err := client.NewDocIDFromString(n.currentValue.GetID())
+			if err != nil {
+				return false, err
+			}
+			doc, err := n.collection.Get(n.p.ctx, docID, false)
+			if err != nil {
+				return false, err
+			}
+			for k, v := range n.updateInput {
+				if err := doc.Set(k, v); err != nil {
+					return false, err
+				}
+			}
+			err = n.collection.Update(n.p.ctx, doc)
+			if err != nil {
+				return false, err
+			}
+		case false:
+			err := n.collection.Create(n.p.ctx, n.doc)
+			if err != nil {
+				return false, err
+			}
+			n.results.Spans(docIDsToSpans(documentsToDocIDs(n.doc), n.collection.Description()))
+		}
+		err = n.results.Init()
+		if err != nil {
+			return false, err
+		}
+		n.isUpserting = false
+	}
+
+	next, err := n.results.Next()
+	if err != nil {
+		return false, err
+	}
+	if !next {
+		return false, nil
+	}
+	n.currentValue = n.results.Value()
+	return true, nil
+}
+
+func (n *upsertNode) Kind() string {
+	return "upsertNode"
+}
+
+func (n *upsertNode) Spans(spans core.Spans) {
+	n.results.Spans(spans)
+}
+
+func (n *upsertNode) Init() error {
+	return n.results.Init()
+}
+
+func (n *upsertNode) Start() error {
+	doc, err := client.NewDocFromMap(n.createInput, n.collection.Definition())
+	if err != nil {
+		return err
+	}
+	n.doc = doc
+
+	return n.results.Start()
+}
+
+func (n *upsertNode) Close() error {
+	return n.results.Close()
+}
+
+func (n *upsertNode) Source() planNode {
+	return n.results
+}
+
+func (n *upsertNode) simpleExplain() (map[string]any, error) {
+	simpleExplainMap := map[string]any{}
+
+	// Add the filter attribute if it exists, otherwise have it nil.
+	if n.filter == nil {
+		simpleExplainMap[filterLabel] = nil
```

Comment:
question: is it legal to have no filter?

todo: regardless might be an easy test to add without a filter in an explain test to hit this line

Reply:
A filter is required. I've removed the nil check.